File: src/menu/components/Panel.java
package menu.components;

import java.awt.Color;

import core.Screen;
import menu.ComponentContainer;
import menu.MenuComponent;
import menu.Menu;
import misc.math.menu.MenuRect;
import misc.math.menu.MenuSize;

public class Panel extends ComponentContainer
{
	private static final MenuSize padding = new MenuSize(20, 20); // über "padding" wird der Abstand innerhalb des KomponentenContainers zu den Komponenten gespeichert. Wird in calcMenuSize() benötigt

	public Panel(ComponentContainer parent, MenuRect rect)
	{
		super(parent, rect);
	}

	@Override public void render()
	{
		super.render(); // rendere alle MenuComponents
		Screen.g().setColor(Color.BLUE);
		Screen.g().drawRect(getOffset().getX(), getOffset().getY(), getWidth()-1, getHeight()-1); // zeichne blauen rand
	}

	// ändert die MenuSize des Panels so, dass zwischen der Komponente links unten noch padding Platz ist
	public void calcMenuSize()
	{
		int right = 0;
		int bot = 0;

		// sucht die width/height indem alle Komponenten durchlaufen werden und die am weitesten rechts und unten gespeichert werden
		for (MenuComponent component : getComponents())
		{
			right = Math.max(right, component.getRight() + padding.getX());
			bot = Math.max(bot, component.getBot() + padding.getY());
		}

		if (right != this.getWidth())
		{
			this.setWidth(right);
		}

		if (bot != this.getHeight())
		{
			this.setHeight(bot);
		}
	}
}

File: cpp/src/menu/components/Panel.cpp
#include "Panel.hpp"

#include <algorithm>

#include <core/Screen.hpp>

Panel::Panel(ComponentContainer* c, const PixelRect& r)
	: MenuComponent(c, r)
{}

PixelRect Panel::getAbsoluteRect() const
{
	return MenuComponent::getAbsoluteRect();
}

PixelRect Panel::getRelativeRect() const
{
	return MenuComponent::getRelativeRect();
}

MenuComponent* Panel::getHoveredComponentRecursively() const
{
	return ComponentContainer::getHoveredComponent();
}

void Panel::render() const
{
	Screen::fillRect(getAbsoluteRect(), sf::Color(255, 255, 255, 40));
	ComponentContainer::render();
}

void Panel::calcSize()
{
	int width = 0;
	int height = 0;

	// sucht die width/height indem alle Komponenten durchlaufen werden und die am weitesten rechts und unten gespeichert werden
	for (unsigned int i = 0; i < getComponents().size(); i++)
	{
		width = std::max(width, getComponents()[i]->getRelativeRect().getRight() + getPadding().x);
		height = std::max(height, getComponents()[i]->getRelativeRect().getBot() + getPadding().y);
	}

	if (width != getRelativeRect().getSize().x)
	{
		setRelativeRect(PixelRect(
			getRelativeRect().getPosition().x,
			getRelativeRect().getPosition().y,
			width,
			getRelativeRect().getSize().y));
	}

	if (height != getRelativeRect().getSize().y)
	{
		setRelativeRect(PixelRect(
			getRelativeRect().getPosition().x,
			getRelativeRect().getPosition().y,
			getRelativeRect().getSize().x,
			height));
	}
}

PixelVector Panel::getPadding() const { return PixelVector(20, 20); }
